src/compute/src/logging/persist.rs

```diff
@@ -7,68 +7,55 @@
 // the Business Source License, use of this software will be governed
 // by the Apache License, Version 2.0.
 
-use std::any::Any;
-use std::cell::RefCell;
-use std::rc::Rc;
 use std::sync::Arc;
 
-use differential_dataflow::AsCollection;
-use differential_dataflow::Collection;
-use differential_dataflow::Hashable;
-use timely::dataflow::channels::pact::Exchange;
-use timely::dataflow::operators::generic::builder_rc::OperatorBuilder;
+use differential_dataflow::consolidation::consolidate_updates;
+use differential_dataflow::input::Input;
+use differential_dataflow::{Collection, Hashable};
 use timely::dataflow::Scope;
 use timely::progress::Antichain;
-use timely::progress::Timestamp as TimelyTimestamp;
-use timely::PartialOrder;
+use tokio::runtime::Handle as TokioHandle;
+use tokio::sync::Mutex;
 
+use mz_persist_client::cache::PersistClientCache;
 use mz_repr::GlobalId;
 use mz_repr::{Diff, Row, Timestamp};
 use mz_storage::controller::CollectionMetadata;
-use mz_storage::types::errors::DataflowError;
-use mz_storage::types::sinks::SinkAsOf;
 use mz_storage::types::sources::SourceData;
-use mz_timely_util::operators_async_ext::OperatorBuilderExt;
 
 use crate::compute_state::ComputeState;
 
 pub(crate) fn persist_sink<G>(
     target_id: GlobalId,
     target: &CollectionMetadata,
     compute_state: &mut ComputeState,
-    desired_collection: &Collection<G, Row, Diff>,
+    desired_collection: Collection<G, Row, Diff>,
 ) where
     G: Scope<Timestamp = Timestamp>,
 {
+    // The target storage collection might still contain data written by a
+    // previous incarnation of the replica. Empty it, so we don't end up with
+    // stale data that never gets retracted.
+    // TODO(teskje,lh): Remove the truncation step once/if #13740 gets merged.
+    //
+    // We need to ensure that only a single timely worker tries to perform the
+    // truncate, to avoid retracting the old data multiple times.
     let scope = desired_collection.scope();
-    let as_of: SinkAsOf = SinkAsOf {
-        frontier: Antichain::from_elem(0),
-        strict: false,
-    };
-
-    // To create the diffs we also need to read from persist.
-    let (ok_stream, err_stream, token) = mz_storage::source::persist_source::persist_source(
-        &scope,
-        Arc::clone(&compute_state.persist_clients),
-        target.clone(),
-        as_of.frontier.clone(),
+    let is_active_worker = (target_id.hashed() as usize) % scope.peers() == scope.index();
+    if is_active_worker {
+        let persist_clients = Arc::clone(&compute_state.persist_clients);
+        TokioHandle::current().block_on(truncate_storage_collection(target, persist_clients));
```

Comment:
One comment I got for the old PR was that the operator creation should be fast, I think this blocks on the snapshot fetch. I wonder if the truncate should be part of the persist_sink API, then we could flush it in the async operator

Reply:
I though about that too. My assumption was kind of that https://github.com/MaterializeInc/materialize/pull/13740 would be production ready soon and that we really only need a quick workaround that is good enough for a couple weeks.

But I think it wouldn't actually be too much effort to move this into the `persist_sink` operator and add a `truncate: bool` flag to switch it on/off. I'll look into that later, if there are no other concerns with this solution.

Reply:
I think the snapshot fetch is more acceptable than the 100% CPU. So my take is, we should merge it as is. And then debug the old solution (I plan to start working on it later today)